maas/client/viscera/boot_resources.py

```diff
@@ -0,0 +1,123 @@
+"""Objects for boot resources."""
+
+__all__ = [
+    "BootResource",
+    "BootResources",
+]
+
+from typing import Mapping
+
+from . import (
+    check,
+    check_optional,
+    Object,
+    ObjectField,
+    ObjectSet,
+    ObjectType,
+    parse_timestamp,
+)
+
+
+class BootResourceMappingType(ObjectType):
+    """Metaclass for `BootResourceSet` and `BootResourceFile`."""
+
+    def mapping(cls, data):
+        """Convert `data` mapping to mapping for class."""
+        if not isinstance(data, Mapping):
+            raise TypeError(
+                "data must be a mapping, not %s"
+                % type(data).__name__)
+        return {
+            key: cls(value)
+            for key, value in data.items()
+        }
+
+
+class BootResourceFile(Object, metaclass=BootResourceMappingType):
+    """A boot resource file."""
+
+    filename = ObjectField.Checked(
+        "filename", check(str), readonly=True)
+    filetype = ObjectField.Checked(
+        "filetype", check(str), readonly=True)
+    size = ObjectField.Checked(
+        "size", check(int), readonly=True)
+    sha256 = ObjectField.Checked(
+        "sha256", check(str), readonly=True)
+    complete = ObjectField.Checked(
+        "complete", check(bool), readonly=True)
+
+
+class BootResourceSet(Object, metaclass=BootResourceMappingType):
+    """A boot resource set."""
+
+    version = ObjectField.Checked(
+        "version", check(str), readonly=True)
+    size = ObjectField.Checked(
+        "size", check(int), readonly=True)
+    label = ObjectField.Checked(
+        "label", check(str), readonly=True)
+    complete = ObjectField.Checked(
+        "complete", check(bool), readonly=True)
+    files = ObjectField.Checked(
+        "files", BootResourceFile.mapping, default={})
+
+
+class BootResourcesType(ObjectType):
+    """Metaclass for `BootResources`."""
+
+    def __iter__(cls):
+        return map(cls._object, cls._handler.read())
+
+
+class BootResources(ObjectSet, metaclass=BootResourcesType):
+    """The set of boot resources."""
+
+    @classmethod
+    def read(cls, *, full=False):
+        """Get list of `BootResource`'s.
+
+        :param full: Load all content on the read. Default `sets` are not
+            loaded.
+        :type full: bool
+        """
+        if full:
+            return cls([
+                cls._object(cls._object._handler.read(id=data['id']))
+                for data in cls._handler.read()
+            ])
```

Comment:
This is hiding an O(n) operation from the caller and should be addressed on the server, not here. In the meantime a workaround can be documented in the docstring. An implementation of that workaround should not be in this codebase.